File: iosApp/deadly/Feature/Notifications/NotificationBody.swift
import SwiftUI

// Render a notification body with tappable links (decision D). Supports
// markdown links `[label](url)` AND bare http(s) URLs, restricted to the
// http/https schemes. Built as an AttributedString with `.link` runs so
// SwiftUI Text opens them in Safari on tap — no HTML, no manual gestures.

private let notificationLinkRegex = try! NSRegularExpression(
    pattern: #"\[([^\]]+)\]\((https?://[^\s)]+)\)|(https?://[^\s]+)"#
)

private let trailingPunctuation = CharacterSet(charactersIn: ").,!?;:")

func notificationBody(_ body: String, linkColor: Color) -> AttributedString {
    var result = AttributedString("")
    let ns = body as NSString
    var last = 0

    func link(_ urlString: String, label: String) -> AttributedString {
        var run = AttributedString(label)
        if let url = URL(string: urlString), url.scheme == "http" || url.scheme == "https" {
            run.link = url
            run.foregroundColor = linkColor
            run.underlineStyle = .single
        }
        return run
    }

    for m in notificationLinkRegex.matches(in: body, range: NSRange(location: 0, length: ns.length)) {
        if m.range.location > last {
            result += AttributedString(ns.substring(with: NSRange(location: last, length: m.range.location - last)))
        }
        let mdUrlRange = m.range(at: 2)
        let bareRange = m.range(at: 3)
        if mdUrlRange.location != NSNotFound {
            result += link(ns.substring(with: mdUrlRange), label: ns.substring(with: m.range(at: 1)))
        } else if bareRange.location != NSNotFound {
            var urlStr = ns.substring(with: bareRange)
            var trailing = ""
            while let lastChar = urlStr.unicodeScalars.last, trailingPunctuation.contains(lastChar) {
                trailing = String(lastChar) + trailing
                urlStr.removeLast()
            }
            result += link(urlStr, label: urlStr)
            if !trailing.isEmpty { result += AttributedString(trailing) }
        }
        last = m.range.location + m.range.length
    }
    if last < ns.length {
        result += AttributedString(ns.substring(from: last))
    }
    return result
}

File: androidApp/app/src/main/java/com/grateful/deadly/notifications/NotificationBody.kt
package com.grateful.deadly.notifications

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.LinkAnnotation
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextLinkStyles
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.withLink
import androidx.compose.ui.text.style.TextDecoration

// Render a notification body with tappable links (decision D). Supports
// markdown links `[label](url)` AND bare http(s) URLs, restricted to the
// http/https schemes. Built as an AnnotatedString with LinkAnnotation.Url so
// Compose Text handles taps (opening the system browser) — no manual gesture
// wiring, and no HTML injection surface.

// `[label](url)` OR a bare http(s):// URL. Markdown alt first so a URL already
// inside a markdown link isn't double-matched.
private val TOKEN =
    Regex("""\[([^\]]+)\]\((https?://[^\s)]+)\)|(https?://[^\s]+)""")

private val TRAILING = Regex("""[).,!?;:]+$""")

fun notificationBody(body: String, linkColor: Color): AnnotatedString =
    buildAnnotatedString {
        val styles = TextLinkStyles(
            style = SpanStyle(color = linkColor, textDecoration = TextDecoration.Underline),
        )
        var last = 0
        for (match in TOKEN.findAll(body)) {
            if (match.range.first > last) append(body.substring(last, match.range.first))
            val mdLabel = match.groupValues[1]
            val mdUrl = match.groupValues[2]
            val bareUrl = match.groupValues[3]
            when {
                mdUrl.isNotEmpty() -> {
                    withLink(LinkAnnotation.Url(mdUrl, styles)) { append(mdLabel) }
                }
                bareUrl.isNotEmpty() -> {
                    val trailing = TRAILING.find(bareUrl)?.value ?: ""
                    val url = bareUrl.dropLast(trailing.length)
                    withLink(LinkAnnotation.Url(url, styles)) { append(url) }
                    if (trailing.isNotEmpty()) append(trailing)
                }
                else -> append(match.value)
            }
            last = match.range.last + 1
        }
        if (last < body.length) append(body.substring(last))
    }
